sql/catalyst/src/main/scala/org/apache/spark/sql/util/SQLOpenHashSet.scala

```diff
@@ -85,12 +85,13 @@ object SQLOpenHashSet {
       index: String,
       hashSet: String,
       handleNotNull: (String, String) => String,
-      handleNull: String): String = {
+      handleNull: String,
+      nullConditions: Seq[String] = Seq.empty): String = {
     if (arrayContainsNull) {
       if (setContainsNull) {
         s"""
            |if ($array.isNullAt($index)) {
-           |  if (!$hashSet.containsNull()) {
+           |  if (!$hashSet.containsNull()${nullConditions.map(" &&" + _).mkString("")}) {
```

Comment:
Why is the implementation different from interpreted path? e.g., `SQLOpenHashSet.withNullCheckFunc`

Reply:
same question

Reply:
Caused by `ArrayType`'s `containsNull` is not reliable.  
When use `Literal.create(Seq(1d), ArrayType(DoubleType))` this array's `containsNull` is also true

Reply:
The whole stage codegen's code seems wrong ....in origins code it's ok seems it use a flag foundNullElement to flag if array2 contains null.

```
        def withArray1NullAssignment(body: String) =
          if (left.dataType.asInstanceOf[ArrayType].containsNull) {
            if (right.dataType.asInstanceOf[ArrayType].containsNull) {
              s"""
                 |if ($array1.isNullAt($i)) {
                 |  if ($foundNullElement) {
                 |    $nullElementIndex = $size;
                 |    $foundNullElement = false;
                 |    $size++;
                 |    $builder.$$plus$$eq($nullValueHolder);
                 |  }
                 |} else {
                 |  $body
                 |}
               """.stripMargin
            } else {
              s"""
                 |if (!$array1.isNullAt($i)) {
                 |  $body
                 |}
               """.stripMargin
            }
          } else {
            body
          }
```

Reply:
How about current's code path?

Reply:
> Why is the implementation different from interpreted path? e.g., `SQLOpenHashSet.withNullCheckFunc`

Their own code generation methods are different. There will be containsNull related judgments in codegen, so the extraction methods are different. The problem this time is caused by the surface problem of the wrong understanding of the field.